autograd/profiler: make python record_function use JIT methods

d4l3k opened the pull request:
This makes the python autograd.profiler.record_function use two new jit methods. This allows for using record_function in TorchScript modules.

See #28249 for discussion.

Differential Revision: D17997612

bddppq commented on test/test_autograd.py:
```diff
@@ -2634,15 +2634,23 @@ def test_record_function(self):
+            'profiler::record_function_exit',
+            'add',
+        ]
+        print(p.function_events)
```
remove this?

d4l3k replied:
removed

bddppq reviewed (commented):
Nice, this looks good to me.

bddppq commented on test/test_autograd.py:
```diff
@@ -2634,15 +2634,23 @@ def test_record_function(self):
+            self.assertGreater(info.cpu_interval.start, last_start)
             self.assertEqual(info.name, expected_name)
-            last_end = info.cpu_interval.end
+            last_start = info.cpu_interval.start
```
Shall we also check "label" event's time range is greater than "mul" and "add"?

d4l3k replied:
added a check for the end ordering

d4l3k reviewed (commented):
I updated this so the thread_local_func_ is always set correctly and added some asserts to ensure that is in fact the case.

bddppq commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -189,14 +196,42 @@ void RecordFunction::processCallbacks() {
 
+at::Tensor record_function_enter(const std::string& name) {
+  if (thread_local_func_) {
+    thread_local_func_->end();
```
Oh good catch! Could you:

1. (nit) assert the `thread_local_func_`'s name is indeed "record_function_enter"
2. add a comment explaining this (it took me quite a while to figure out why we need to do this :-))

d4l3k replied:
done

albanD reviewed (approved):
I'm not very familiar with the profiler code :/

But the diff looks good to me.

bddppq commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -189,14 +196,42 @@ void RecordFunction::processCallbacks() {
+void record_function_exit(at::Tensor handle) {
+  // We don't actually need to do anything with handle just need to persist the
+  // lifetime until now.
+  auto& rec = at::cpp_custom_type_hack::cast<RecordFunction>(handle);
```
Actually I'm a little bit confused here, how is the `RecordFunction::end` (or `~RecordFunction`) getting triggered?

albanD replied:
My understanding was that this was linked to the `handle` object.

d4l3k replied:
It was getting triggered when the Tensor was freed after record_function_exit finished. It's a bit confusing and now that we have end() I modified it to explicitly call it to make it clearer.

d4l3k commented:
Updated diff w/ comments + added a test for when the profiler is not enabled but the ops are called anyways.

d4l3k commented:
fixed lint

bddppq commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -189,14 +193,60 @@ void RecordFunction::processCallbacks() {
+    AT_ASSERT(
+        thread_local_func_->name() ==
+        StringView("profiler::record_function_exit"));
+    thread_local_func_->end();
```
nit: I'm wondering if it's better to have the scope event ended earlier than record_function_exit?

d4l3k replied:
You mean have the scope end before record_function_exit runs? It's pretty hard to do that with the current thread_local_func & parent_ stack.

bddppq reviewed (approved):
LGTM

bddppq commented:
Tests failed on windows:
```
19:27:36 ======================================================================
19:27:36 FAIL: test_record_function (__main__.TestAutograd)
19:27:36 ----------------------------------------------------------------------
19:27:36 Traceback (most recent call last):
19:27:36   File "test_autograd.py", line 2678, in test_record_function
19:27:36     self.assertEqual(info.name, expected_name)
19:27:36   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\common_utils.py", line 738, in assertEqual
19:27:36     super(TestCase, self).assertEqual(x, y, message)
19:27:36 AssertionError: 'inner' != 'profiler::record_function_exit'
19:27:36 - inner
19:27:36 + profiler::record_function_exit
19:27:36  : 
```
I'm wondering if it's the end time of these events happen to be the same in this run so their order become nondeterministic after sorting.

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -189,14 +193,60 @@ void RecordFunction::processCallbacks() {
+  return at::cpp_custom_type_hack::create(std::move(rec), at::TensorOptions());
+}
+
+void record_function_exit(const at::Tensor& handle) {
```
just from code organization perspective - move it to separate file (like record_function_ops)? we might move profiler around in the future and it's better not to create circular dependencies

d4l3k replied:
Unfortunately these ops need to access thread_local_func_ which is only available in this file. @dzhulgakov  Thoughts on how best to resolve that?

bddppq replied:
@d4l3k Make a declaration in torch/csrc/autograd/record_function.h?

d4l3k replied:
Ended up adding a static RecordFunction::current() method to return thread_local_func_. Don't really want to expose direct access to it since it's somewhat fragile.

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -189,14 +193,60 @@ void RecordFunction::processCallbacks() {
+
+static auto registry =
+    RegisterOperators()
+        .op("profiler::record_function_enter", &record_function_enter)
```
can you change the name to be with underscore for now as it's not official API yet? Like `profiler::_record_functiion_enter`? we probably would want to expose nice `with` statements eventually as the recommended way of doing it

d4l3k replied:
we do expose nice `with` statements in this diff :)

dzhulgakov reviewed (commented):
Nice! Minor code organization comment

d4l3k committed:
autograd/profiler: make python record_function use JIT methods

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/28264

Test Plan: buck test caffe2/test:autograd caffe2/test/cpp/jit:jit

Reviewed By: bddppq

Differential Revision: D17997612

fbshipit-source-id: 4a8c66d44f996d4a72b5f29bbbef2bd5b366afc2

bddppq reviewed (approved):
LG! Thanks
